src/serac/numerics/functional/tuple_tensor_dual_functions.hpp

```diff
@@ -665,4 +676,39 @@ auto solve_scalar_equation(function&& f, double x0, double tolerance, double low
   }
 }
 
+/**
+ * @brief Finds a root of a vector-valued nonlinear function
+ *
+ * Uses Newton-Raphson iteration.
+ *
+ * @tparam function Type for the functor object
+ * @tparam n Vector dimension of the equation
+ * @param f A callable representing the function of which a root is sought. Must take an n-vector
+ * argument and return an n-vector
+ * @param x0 Initial guess for root. Must be an n-vector.
+ * @return A root of @p f.
+ */
+template <typename function, int n>
+auto find_root(function&& f, tensor<double, n> x0)
+{
+  static_assert(std::is_same_v<decltype(f(x0)), tensor<double, n>>,
+                "error: f(x) must have the same number of equations as unknowns");
+
+  double epsilon        = 1.0e-8;
+  int    max_iterations = 10;
+
+  auto x = x0;
+
+  int k = 0;
+  while (k++ < max_iterations) {
```

Comment:
Is there any reason this can't be a for loop? I think it would be more readable.

Reply:
Just changed it. Thanks, that was a bit odd.